[executorch] Populate cadence cpu ops

zonglinpeng opened the pull request:
### Summary
quantized ops in CPU flow are not fully migrated. Adding them all in this PR
#### Custom ops
- quantized_linear_per_tensor_out
- im2row_out
- quantized_convolution_per_tensor_out

#### Native ops
- op: transpose_copy.int_out
- op: eq.Scalar_out
- op: logical_not.out
- op: any.out
- op: native_group_norm.out
- op: sum.IntList_out
- op: select_copy.int_out

### Test plan
python3 -m examples.cadence.operators.quantized_linear_op
python3 -m examples.cadence.models.babyllama

mcremon-meta commented on backends/cadence/reference/operators/operators.h:
```diff
@@ -0,0 +1,59 @@
+
+
+#define ET_FORALL_CADENCE_QUANTIZED_TYPES(_) \
+  _(uint8_t, Byte)                          \
+  _(int8_t, Char)
```
general comment, we should be able to add int16_t and uint16_t already, but let's ignore for now

mcremon-meta commented on backends/cadence/reference/operators/quantized_ops.h:
```diff
@@ -0,0 +1,233 @@
+// scalar.
+#define DECLARE_POINTWISE_TENSOR_QUANTIZED_BINARY_OP(BINARY_FUNC_NAME, OP)    \
+  template <typename T>                                                       \
+  void BINARY_FUNC_NAME(                                                      \
+      const ::executorch::aten::Tensor& X,                                    \
+      float X_scale,                                                          \
+      int32_t X_zero_point,                                                   \
+      const ::executorch::aten::Tensor& Y,                                    \
+      float Y_scale,                                                          \
+      int32_t Y_zero_point,                                                   \
+      float out_scale,                                                        \
+      int32_t out_zero_point,                                                 \
+      ::executorch::aten::Tensor& out) {                                      \
+    const T* __restrict__ X_data = X.const_data_ptr<T>();                     \
+    const T* __restrict__ Y_data = Y.const_data_ptr<T>();                     \
+    T* __restrict__ out_data = out.mutable_data_ptr<T>();                     \
+    size_t Y_numel = Y.numel();                                               \
+    size_t X_numel = X.numel();                                               \
+    float inv_out_scale = 1.0f / out_scale;                                   \
+    /* Tensor that has the same element of X */                               \
+    if (Y_numel == X_numel) {                                                 \
+      for (size_t i = 0; i < X_numel; ++i) {                                  \
+        float x = kernels::dequantize<T>(X_data[i], X_scale, X_zero_point);   \
+        float y = kernels::dequantize<T>(Y_data[i], Y_scale, Y_zero_point);   \
+        float z = x OP y;                                                     \
+        out_data[i] = kernels::quantize<T>(z, inv_out_scale, out_zero_point); \
+      }                                                                       \
+    } /* if Y is a scalar Tensor */                                           \
+    else if (Y_numel == 1) {                                                  \
+      float y = kernels::dequantize<T>(Y_data[0], Y_scale, Y_zero_point);     \
+      for (size_t i = 0; i < X_numel; ++i) {                                  \
+        float x = kernels::dequantize<T>(X_data[i], X_scale, X_zero_point);   \
+        float z = x OP y;                                                     \
+        out_data[i] = kernels::quantize<T>(z, inv_out_scale, out_zero_point); \
+      }                                                                       \
+    } /* other broadcasting cases */                                          \
+    else {                                                                    \
+      ET_DCHECK_MSG(false, "Unsupported broadcasting");                       \
+    }                                                                         \
+  }
```
this macro is not used. It's only for quantized_add, and will need to be removed anyway with PT2 quant

mcremon-meta reviewed (approved):
LGTM with the extra macro removed. Thanks!

Commit:
create quantized_linear_per_tensor_out in cpu

Commit:
Populate cadence cpu ops (#7165)

Summary:
quantized ops in CPU flow are not fully migrated. Adding them all in this PR
- quantized_linear_per_tensor_out

Test Plan: python3 -m examples.cadence.operators.quantized_linear_op

Reviewed By: hsharma35, mcremon-meta

Differential Revision: D66726864

Pulled By: zonglinpeng

Commit:
Populate cadence cpu ops (#7165)

Summary:
quantized ops in CPU flow are not fully migrated. Adding them all in this PR
- quantized_linear_per_tensor_out


Test Plan: python3 -m examples.cadence.operators.quantized_linear_op

Reviewed By: hsharma35, mcremon-meta

Differential Revision: D66726864

Pulled By: zonglinpeng

Commit:
removed unused headers and lint

hietalajulius committed:
Feat/Add a React Native LLaMA demo app for iOS 

Differential Revision: D66644092

Pull Request resolved: https://github.com/pytorch/executorch/pull/7134

Commit:
add missing aten ops for x86 flow

Commit:
remove unused macro

mcremon-meta commented on .lintrunner.toml:
```diff
@@ -74,6 +74,7 @@ exclude_patterns = [
     # NB: Objective-C is not supported
     'examples/apple/**',
     'examples/demo-apps/apple_ios/**',
+    'examples/demo-apps/react-native/rnllama/ios/**',
```
lint change I guess?

zonglinpeng replied:
caused by rebasing to main

mcremon-meta commented on backends/cadence/reference/operators/quantized_conv_out.cpp:
```diff
@@ -1,261 +0,0 @@
- * LICENSE file in the root directory of this source tree.
- */
-
-#include <executorch/backends/cadence/reference/kernels/kernels.h>
```
can we put the `convolution.cpp` stuff in this one instead, and remove the non-quant version? It should always be quantized anyway, and it will be less confusing

mcremon-meta reviewed (commented):
not sure why there are so many changes, maybe a linter error?

Commit:
miss 1 rebase

Commit:
rename quant con
